eox_nelp/notifications/tests/test_notify_course_due_date.py

```diff
@@ -0,0 +1,138 @@
+    @patch("eox_nelp.notifications.notify_course_due_date.CourseEmailTemplate")
+    @patch("eox_nelp.notifications.notify_course_due_date.get_xblock_course_from_usage_key_string")
+    @patch("eox_nelp.notifications.notify_course_due_date.get_xblock_from_usage_key_string")
+    @patch("eox_nelp.notifications.notify_course_due_date.CourseEnrollment")
+    def test_notify_upcoming_course_due_date(  # pylint: disable=too-many-arguments
+        self,
+        course_enrollment_mock,
+        get_xblock_from_usage_key_string_mock,
+        get_xblock_course_from_usage_key_string_mock,
+        course_email_template_mock,
+        send_email_multialternative_mock,
+    ):
+        """Test correct behaviour of sending email with
+        method `notify_upcoming_course_due_date`.
+        Expected behavior:
+            - email_key_words are presented in the `text_message` arg (0) of 2nd call of course_email_template.
+            - email_key_words are presented in the `html_message` arg (0) of 3rd call of course_email_template.
+            - course_name is presented in `subject` arg(0) of the call of send_email_multialternative.
+            - subsection_name is presented in `subject` arg(0) of the call of send_email_multialternative.
+            - recipient list is presented in `recipient_emails` arg(3) of the call of send_email_multialternative.
+            - The email was sent from the correct email address (if specified)
+            - sent attribute of instance of `UpcomingCourseDueDate` after the process is True.
+            - Log successful of sent message.
+        """
+        UpcomingCourseDueDate.objects.all().delete()  # pylint: disable=no-member
+        instance, _ = UpcomingCourseDueDate.objects.get_or_create(  # pylint: disable=no-member
+            course=CourseOverview.objects.create(id="course-v1:test+Cx105+2022_T4"),
+            location_id="block-v1:test+CS501+2022_T4+type@sequential+block@a54730a9b89f420a8d0343dd581b447a",
+            due_date=self.subsection_due_date,
+            notification_date=timezone.now(),
+        )
+        sent_emails = 1
+        course_enrollment_mock.objects.filter.return_value = generate_list_mock_data(self.enrollments_data)
+        get_xblock_from_usage_key_string_mock.return_value = generate_list_mock_data(self.subsection_xblock_data)[0]
+        get_xblock_course_from_usage_key_string_mock.return_value = generate_list_mock_data(self.course_xblock_data)[0]
+        send_email_multialternative_mock.return_value = sent_emails
+        email_key_words = [
+            self.course_name,
+            self.subsection_name,
+            str(self.subsection_due_date),
+        ]
+
+        with self.assertLogs(notify_course_due_date.__name__, level="INFO") as logs:
+            notify_upcoming_course_due_date(instance)
+
+        for word in email_key_words:
+            self.assertIn(word, course_email_template_mock.mock_calls[1][1][0])
```

Comment:
why are you testing in this way that the word is in `text_message` when you can use an `assert_called_with ` and test the whole message instead of just one word ?

Reply:
I cant use `assert_called_with` because it use the last call args.  Maybe I can change with the  `assert_any_call` https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_any_call.
No wy the problem is that I cant generate all the message because there is a time now generated, in term of seconds would change.  https://github.com/eduNEXT/eox-nelp/blob/fda7e2bc2e4ffc5055707b38b6c56845a3f515c2/eox_nelp/notifications/notify_course_due_date.py#L48

Reply:
https://github.com/eduNEXT/eox-nelp/blob/fda7e2bc2e4ffc5055707b38b6c56845a3f515c2/eox_nelp/notifications/notify_course_due_date.py#L56

Reply:
so your method is calling this multiple times ?

Reply:
yes, the course_email_template_mock yes. But the specific method not. Let me check if I could tried with that.
https://github.com/eduNEXT/eox-nelp/blob/fda7e2bc2e4ffc5055707b38b6c56845a3f515c2/eox_nelp/notifications/notify_course_due_date.py#L67-L68

Reply:
But the problem is the same, I cant replicate `text_message` or `html_message` due delta calculated when is called  the function.

Reply:
so there are something wrong with your implementation base on your code `CourseEmailTemplate` is never called, the attribute `get_template` is called, probably that returns  a mock that  has  other mock attributes like `render_plaintext`

Your time issue don't make you wonder why are you sending time with milliseconds ? in this context is that worthy?

Reply:
it's worthy it If  I build a string with that. So assert_called_with will fail.
**https://github.com/eduNEXT/eox-nelp/blob/fda7e2bc2e4ffc5055707b38b6c56845a3f515c2/eox_nelp/notifications/notify_course_due_date.py#L60**

Reply:
Is would be different. 
![image](https://user-images.githubusercontent.com/51926076/233738807-aa105395-dbf5-49da-96ac-addc450b508a.png)

Reply:
same comment, in this context is that worthy? do you think the user will find that information useful? why do you think  the value  194313 is important in this context ?